### PR TITLE
`Discussion`: Improve announcement post handling and formatting

### PR DESCRIPTION
**Pull request by lschlesinger**
#### General

- [ ] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.ase.in.tum.de/dev/guidelines/language-guidelines/).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://artemis-platform.readthedocs.io/en/latest/dev/guidelines/development-process.html#naming-conventions-for-github-pull-requests).
#### Server
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/server/).
- [x] I implemented the changes with a good performance and prevented too many database calls.
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] I followed the [coding and design guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client/).
- [x] I added multiple integration tests (Jest) related to the features (with a high test coverage), while following the [test guidelines](https://docs.artemis.ase.in.tum.de/dev/guidelines/client-tests/).
- [x] I documented the TypeScript code using JSDoc style.
- [x] I added multiple screenshots/screencasts of my UI changes.
- [x] I translated all newly inserted strings into English and German.

### Motivation and Context

This PR addresses several shortcomings of the current implementation and enhances some announcement/notification related features:
- Answer post notification in the notification side-bar showed author of original post as "author" and not the answer post that is notified about. This was misleading.
- Info banner on announcement creation is improved: instructor is informed about automatic email notification
- Email notification on announcement changed format -> title and course are in subject (Betreff) instead of "New announcement".
- Content in email was not shown as html rendered markdown but rather raw markdown -> formatting of original post content was lost in email, now it's the same format.
- Code block and reference were not rendered correctly anymore in post content due to a markdown issue introduced earlier
- Tutors could edit an announcement, but not create -> now they are  _not_ allowed to edit or create an announcement

### Description

I addressed the shortcomings/enhancements mentioned above. Most of same imply just minor code changes.
However, I had to add a new pipe `htmlForPostingMarkdown` that mainly relies on the same logic behind the existing `htmlForMarkdown` pipe.
For the markdown used in posting content we are facing an additional challenge which are the integrated post references.
To enable rendering the posting content including a link in between, we split up the content into the parts: `contentBeforeReference`, `reference` and `contentAfterReference`. Each of them can contain multiple paragraphs after converting markdown to html.
In order to assure that the reference is in the same line as the text it is embedded we need to manipulate the paragraph that is directly before the reference and the paragraph that comes directly after the reference. Those should have the css attribute `display: inline` explicitly set.
This was previously achieved in a different way which broke the rendering of code-block markdown and reference markdown.
Now, with the dedicated pipe `htmlForPostingContent` we achieve that the markdown formatting is not broken for code-block and reference _and_ the reference (if there is one) is in the same line as the content around it.

Furthermore, we added the lightweight markdown-to-html paser **[commonmark](https://github.com/commonmark/commonmark-java)** as new server dependency.
Now, we are able to parse the post content (markdown) into html that can be used within the email template on the server-side.
We decided to do this on the server-side as we cannot guarantee security if the html content that is directly used as mail content can be created (and manipulate by users though intercepting the request!) on client-side.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 1 Tutor
- 1 Course with postings enabled

**TEST ON TS1 to receive email notifications**

1. Log in to Artemis
2. Navigate to Course Discussion
3. As Tutor, create an answer post to an existing post (that was created by someone else) -> check the notification, the name shown should be the author of the answer post (not the post)
4. As tutor, find an announcement in the course posts -> check that there is not icon to edit or delete it
5. As tutor, create a new post on the course overview -> check that you cannot select `Announcement` as topic
6. As tutor, find an existing post, that is no announcement post, edit it -> check that you cannot select `Announcement` as topic
7. As instructor, start create a new announcement post (create new post at course overview, select `Announcement` as topic) -> check that the blue info box contains the information on email notification
8. As content use markdown formatting that is supported by the editor (bold, italic, underline, code, codeblock, reference, link) and including a reference (e.g #ID)
9. Save the announcement -> see if the content is rendered correctly
10. Check the email that was sent -> check if it has the title of the announcement as well as the course as subject, -> check if the content of the mail is formatted as seen on the artemis page

#### Code Review
- [ ] Review 1
- [ ] Review 2
#### Manual Tests
- [ ] Test 1
- [ ] Test 2

### Test Coverage

posting-header: 100%
markdown.service.ts: 83.6% (newly added part completely covered by `postings-content-part.component.spec.ts`)
post-header.component.ts: 100%
posting-content-part.component.ts: 100%
html-for-post-markdown.pipe.ts: 100%

Applied Changes to GroupNotificationServiceTest (still ~100%)

Local tests all pass
image

### Screenshots

Email Content Before:
Bildschirmfoto 2021-11-18 um 14 49 45

Email Content After:
image

Markdown Post Content Before:

Bildschirmfoto 2021-11-18 um 14 48 03

Markdown Post Content After:
Bildschirmfoto 2021-11-18 um 14 53 47

### CODE REVIEW

File: src/main/resources/templates/mail/notification/announcementPostEmail.html
Comment (lschlesinger) [inline src/main/resources/templates/mail/notification/announcementPostEmail.html:10]: @Malyuk-A I am not sure how I should integrate strings including html tags here...is `utext` (unescaped text) the correct approach?

**Review by MDK5 — CHANGES_REQUESTED** [review]
Tested in testings session, two issues found:

- As mentioned, the endpoints should be also secured to defend against request forgery
- The URLs in the emails are broken/not existent

**Comment by lschlesinger** [comment]
@MDK5 I fixed it, as can be seen in the updated screenshot of the email content. The links are displayed now and work as expected.
Also I added additional server-side checks for preventing tutors from creating, editing or deleting an announcement. I also added according test cases.

File: src/main/java/de/tum/in/www1/artemis/service/MailService.java
Comment (Olimbia23) [inline src/main/java/de/tum/in/www1/artemis/service/MailService.java:176]: [nit] What do you think of introducing `getExerciseType()` method on the Exercise class and then provide the trivial implementation for it, as in:
```java
enum ExerciseType {
  case TEXT, MODELING, ...
}

class TextExercise {
  ExerciseType getExerciseType() {
    return ExerciseType.TEXT;
  }
}
```

We can make this method (`getExerciseType`) as required one and then if new Exercise is introduced, this code logic (`setExerciseContext`) will just work without any extra though.

I guess otherwise it makes sense to at least log (or assert) in the `else` case, as we won't set the exercise for the exercise which will hit the else case.

PS. Oh, just noticed it's not your code.

File: src/main/java/de/tum/in/www1/artemis/service/MailService.java
Comment (Olimbia23) [inline src/main/java/de/tum/in/www1/artemis/service/MailService.java:226]: [nit] extra newline

File: src/main/java/de/tum/in/www1/artemis/service/MailService.java
Comment (Olimbia23) [inline src/main/java/de/tum/in/www1/artemis/service/MailService.java:195]: [nit] I assume there is no translation-like service on the server, as we have on the client, right?
I think we could use formatted strings instead, in order to make it more readable.
Something like:
```
final String newAnnouncementEN = "New announcement \"%s\" in course \"%s\"".
final String newAnnouncementDE = "...";

...
private String setPostContextAndSubject(Context context, Object notificationSubject, Locale locale) {
  ...
  String newAnnouncement = locale.toString().equals("en") ? newAnnouncementEN : newAnnouncementDE ;
  return String.format(newAnnoucement, 
                       ((Post) notificationSubject).getTitle(), 
                       ((Post) notificationSubject).getCourse().getTitle());
}
```
I guess this would also be more easier to port to the translation-like service in the future.

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (Olimbia23) [inline src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:476]: Should we also add the "ElseThrow" suffix to the method name?

**Review by Olimbia23 — APPROVED** [review]
Nice work @lschlesinger 
Code changes look good.
Approve 👍 

PS. I left some minor comments

**Review by MerlinMehmed — DISMISSED** [review]
Tested on TS1, all test steps worked as described!

File: src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java
Comment (lschlesinger) [reply src/main/java/de/tum/in/www1/artemis/service/metis/PostService.java:476]: yes, it's more consistent, I will change it.

File: src/main/java/de/tum/in/www1/artemis/service/MailService.java
Comment (lschlesinger) [reply src/main/java/de/tum/in/www1/artemis/service/MailService.java:176]: yes, same applies to the other comments. I think it's better to address this in a mail/notification related PR. But thanks for the input :)

File: src/main/java/de/tum/in/www1/artemis/service/MailService.java
Comment (Malyuk-A) [reply src/main/java/de/tum/in/www1/artemis/service/MailService.java:195]: This is quite interesting, I have not yet seen this in Java (it reminds me of C). Thank you very much for this idea.
Because this is an issue of code quality and not related to Lorena's work (explicitly) I will implement this (and your other remark) in a small follow-up PR. The logic works and I think this should not hinder this Metis PR from being merged.

**Review by Olimbia23 — APPROVED** [review]
Thank you for applying my suggestions @lschlesinger 
Code changes look good.
Approve 👍

File: src/main/java/de/tum/in/www1/artemis/service/MailService.java
Comment (Malyuk-A) [reply src/main/java/de/tum/in/www1/artemis/service/MailService.java:176]: Very good idea. I simply crafted something more or less on the fly. I added your change to a follow-up PR (see my answer to your other comment about the MailService for more info)

File: src/main/java/de/tum/in/www1/artemis/service/MailService.java
Comment (Malyuk-A) [reply src/main/java/de/tum/in/www1/artemis/service/MailService.java:195]: This is the follow up PR which I will open once this PR has been merged. https://github.com/ls1intum/Artemis/pull/4345

**Review by Malyuk-A — DISMISSED** [review]
Re-approve (minor changes after last review)

**Review by willergeorg — APPROVED** [review]
Tested on TS1. No error occurred and link seems to be in the email again.

**Review by MerlinMehmed — APPROVED** [review]
Reapprove!

**Review by Olimbia23 — APPROVED** [review]
Reapprove 👍

**Pull request merged**
